kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java

```diff
@@ -56,49 +58,65 @@ public class InBandKeyManager<K, E> implements KeyManager<K> {
     private final Serde<K> kekIdSerde;
     private final Serde<E> edekSerde;
     // TODO cache expiry, with key descruction
-    private final ConcurrentHashMap<K, CompletionStage<KeyContext>> keyContextCache;
+    private final ConcurrentHashMap<K, CompletableFuture<KeyContext>> keyContextCache;
     private final ConcurrentHashMap<RecordHeader, CompletionStage<AesGcmEncryptor>> decryptorCache;
     private final long dekTtlNanos;
     private final int maxEncryptionsPerDek;
 
     public InBandKeyManager(Kms<K, E> kms,
-                            BufferPool bufferPool) {
+                            BufferPool bufferPool,
+                            int maxEncryptionsPerDek) {
         this.kms = kms;
         this.bufferPool = bufferPool;
         this.edekSerde = kms.edekSerde();
         this.kekIdSerde = kms.keyIdSerde();
         this.dekTtlNanos = 5_000_000_000L;
-        this.maxEncryptionsPerDek = 500_000;
+        this.maxEncryptionsPerDek = maxEncryptionsPerDek;
         // TODO This ^^ must be > the maximum size of a batch to avoid an infinite loop
         this.keyContextCache = new ConcurrentHashMap<>();
         this.decryptorCache = new ConcurrentHashMap<>();
     }
 
     private CompletionStage<KeyContext> getKeyContext(K key,
-                                                      Supplier<CompletionStage<KeyContext>> valueSupplier) {
+                                                      Supplier<CompletableFuture<KeyContext>> valueSupplier,
+                                                      int numRecordsToEncrypt) {
         return keyContextCache.compute(key, (k, v) -> {
             if (v == null) {
                 return valueSupplier.get();
                 // TODO what happens if the CS complete exceptionally
                 // TODO what happens if the CS doesn't complete at all in a reasonably time frame?
             }
             else {
+                if (v.isDone()) {
+                    KeyContext keyContext = v.join();
+                    if (keyContext.isExpiredForEncryption(System.nanoTime())
+                            || keyContext.remainingEncryptions() < numRecordsToEncrypt) {
+                        // is it safe to destroy a context just because the limits are exceeded? we decrement the remaining count before doing encryption work
```

Comment:
Need to figure out if the old KeyContext could still be in use while this is being called